src/main/java/io/getunleash/metric/UnleashMetricServiceImpl.java

```diff
@@ -51,11 +68,62 @@ public void countVariant(String toggleName, String variantName) {
 
     private Runnable sendMetrics() {
         return () -> {
-            MetricsBucket metricsBucket = this.currentMetricsBucket;
-            this.currentMetricsBucket = new MetricsBucket();
-            metricsBucket.end();
-            ClientMetrics metrics = new ClientMetrics(unleashConfig, metricsBucket);
-            metricSender.sendMetrics(metrics);
+            if (interval.get() == 0) {
+                MetricsBucket metricsBucket = this.currentMetricsBucket;
+                this.currentMetricsBucket = new MetricsBucket();
+                metricsBucket.end();
+                ClientMetrics metrics = new ClientMetrics(unleashConfig, metricsBucket);
+                int statusCode = metricSender.sendMetrics(metrics);
+                if (statusCode >= 200 && statusCode < 400) {
+                    if (failures.get() > 0) {
+                        interval.set(Integer.max(failures.decrementAndGet(), 0));
+                    }
+                }
+                if (statusCode >= 400) {
+                    handleHttpErrorCodes(statusCode);
+                }
+            } else {
+                interval.decrementAndGet();
+            }
         };
     }
+
+    private void handleHttpErrorCodes(int responseCode) {
+        if (responseCode == 404) {
+            interval.set(maxInterval);
+            failures.incrementAndGet();
+            LOGGER.error(
+                    "Server said that the Metrics receiving endpoint at {} does not exist. Backing off to {} times our poll interval to avoid overloading server",
+                    unleashConfig.getUnleashURLs().getClientMetricsURL(),
+                    maxInterval);
+        } else if (responseCode == 429) {
+            interval.set(Math.min(failures.incrementAndGet(), maxInterval));
+            LOGGER.info(
+                    "Client Metrics was RATE LIMITED for the {}. time. Further backing off. Current backoff at {} times our metrics post interval",
+                    failures.get(),
+                    interval.get());
+        } else if (responseCode == HttpURLConnection.HTTP_UNAUTHORIZED
+                || responseCode == HttpURLConnection.HTTP_FORBIDDEN) {
+            failures.incrementAndGet();
+            interval.set(maxInterval);
+            LOGGER.error(
+                    "Client was not authorized to post metrics to the Unleash API at {}. Backing off to {} times our poll interval to avoid overloading server",
+                    unleashConfig.getUnleashURLs().getClientMetricsURL(),
+                    maxInterval);
+        } else if (responseCode >= 500) {
+            interval.set(Math.min(failures.incrementAndGet(), maxInterval));
```

Comment:
I'm thinking these modifications of the interval can be added to an internal method like `this.increaseInterval()` or `this.skipMore()` (depending on the name). I believe that can improve the maintainability of the code.

Reply:
Agreed, will extract method